near-sdk-macros/src/core_impl/abi.rs

```diff
@@ -1,51 +1,83 @@
-use crate::core_impl::{utils, AttrSigInfo};
-use crate::core_impl::{BindgenArgType, SerializerType};
-use crate::{ImplItemMethodInfo, MethodType};
 use proc_macro2::{Span, TokenStream as TokenStream2};
-use quote::quote;
+use quote::{format_ident, quote};
 use syn::spanned::Spanned;
-use syn::{ReturnType, Type};
+use syn::{Attribute, Lit::Str, Meta::NameValue, MetaNameValue, ReturnType, Type};
+
+use crate::core_impl::{
+    utils, AttrSigInfo, BindgenArgType, ImplItemMethodInfo, ItemImplInfo, MethodType,
+    SerializerType,
+};
+
+pub fn generate(i: &ItemImplInfo) -> TokenStream2 {
+    let public_functions: Vec<&ImplItemMethodInfo> =
+        i.methods.iter().filter(|m| m.is_public || i.is_trait_impl).collect();
+    if public_functions.is_empty() {
+        // Short-circuit if there are no public functions to export to ABI
+        return TokenStream2::new();
+    }
+
+    let functions: Vec<TokenStream2> = public_functions.iter().map(|m| m.abi_struct()).collect();
+    let first_function_name = &public_functions[0].attr_signature_info.ident;
+    let near_abi_symbol = format_ident!("__near_abi_{}", &first_function_name);
+    quote! {
+        #[cfg(not(target_arch = "wasm32"))]
+        const _: () = {
+            #[no_mangle]
+            pub fn #near_abi_symbol() -> near_sdk::__private::ChunkedAbiEntry {
+                let mut gen = near_sdk::__private::schemars::gen::SchemaGenerator::default();
+                let functions = vec![#(#functions),*];
+                near_sdk::__private::ChunkedAbiEntry::new(functions, gen.into_root_schema_for::<String>())
+            }
+        };
+    }
+}
 
 impl ImplItemMethodInfo {
     /// Generates ABI struct for this function.
     ///
     /// # Example:
     /// The following function:
     /// ```ignore
-    /// fn f3(&mut self, arg0: FancyStruct, arg1: u64) -> Result<IsOk, Error> { }
+    /// /// I am a function.
+    /// pub fn f3(&mut self, arg0: FancyStruct, arg1: u64) -> Result<IsOk, Error> { }
     /// ```
     /// will produce this struct:
     /// ```ignore
     /// near_sdk::__private::AbiFunction {
     ///     name: "f3".to_string(),
+    ///     doc: Some(" I am a function.".to_string()),
```

Comment:
Curious why the initial space isn't removed from these comments. I know unrelated, just wondering if an issue should be  opened up. cc @itegulov

Reply:
The initial space comes from here:
```rust
/// I am a function
   ^
```

I left all original indentation intentionally as it is not clear on what exactly should be deleted. schemars actually tried doing some heuristics here, but they later realized that it does not suit all use cases and can even mess up markdown (see https://github.com/GREsau/schemars/issues/38#issuecomment-826395307). Specifically, here trimming all lines can mess with code indentation in `` ``` `` blocks. So I feel like post-processing should be left to the user depending on their use case as it is easy to remove unwanted stuff but impossible to deduce what stuff has already been removed by us.

Reply:
yeah, my assumption would be that just the first space would be trimmed, don't have opinions of any other line. Seems like that first space will always exist and might be a strange side-effect for those consuming it. Not blocking

Reply:
Hmm, I guess that is fair. I assumed that since Rust keeps this leading space, you could technically do `///this is doc`, but yeah I would not be opposed to trimming it.

Reply:
I'll just open an issue, and we can do it in a separate PR or close it if we don't want to do it. Super minor and opinionated